### ISSUE TITLE
Push participations + retractions to the relay system 

### ISSUE THREAD

**Issue by jaywink** [redmatrix/hubzilla#566]
Hubzilla AFAIK supports pushing posts out to [the relay system](https://github.com/jaywink/social-relay/blob/master/docs/relays.md).

In addition to posts, participations (likes, comments) and retractions will be delivered by the relays to the same destination where the originating post was delivered to. This allows nodes which are fully disconnected from each other by social interaction to receive not only the posts but also the participations.

Hubzilla should, if relay outbound setting is enabled, send also participations (comments, likes) and retractions (for posts) to the relay in the settings. See [relay document](https://github.com/jaywink/social-relay/blob/master/docs/relays.md#participations-comments-likes-and-participations).

Basically TLDR, just like for the post, a simple CC of the payload to the (preferably same) relay.

Please let me know if you have questions about this and sorry for the noise if this is already implemented.

**Comment by git-marijus** [comment]
As far as i can see we already push all public items to relay@relay.iliketoast.net by default...

**Comment by git-marijus** [comment]
I do have a problem though registering my hub to the-federation.info
At least it does not appear in the list of nodes...
My hub address is hub.somaton.com

**Comment by zotlabs** [comment]
Registering a relay is somewhat different than registering for the statistics server. For the latter I believe you need to have the statistics_json addon installed.

**Comment by git-marijus** [comment]
I see... I thought it would show up in stats either way...

Anyway... I have not received any public posts yet  from diaspora...

**Comment by zotlabs** [comment]
We do log the response from the relay subscription, which typically mentions that you've been subscribed. Unfortunately it doesn't seem like a lot of folks are using relays yet. I'm following #diaspora and #hubzilla and get perhaps one post a day. sometimes two.  I've mentioned this to Jason, and it seems that this is about right for current relay usage. I haven't tried following the public firehose; though this should provide a bit more activity.

**Comment by jaywink** [comment]
@git-marijus it looks like hub.somaton.com doesn't offer neither a NodeInfo document or a statistics.json document. One or the other is needed for the-federation.info to be able to pull metadata about your instance.

Probably a Hubzilla addon missing?

![selection_111](https://cloud.githubusercontent.com/assets/1174866/19658589/2f8a09de-9a31-11e6-877a-371de865d830.png)

**Comment by git-marijus** [comment]
Thanks @jaywink!
It seems that the relay server checks for statistics.json to determine if the register url is a valid endpoint. I enabled this plugin now... Hope i will get some public posts now :)

@zotlabs we should probably make the statistics_json plugin a dependency of the diaspora plugin or maybe merge the statistics into diaspora. What do you think?

**Comment by git-marijus** [comment]
nodeinfo has been broken a little... fixed in 4953cdb

**Comment by zotlabs** [comment]
They were mostly kept separate because the statistics publishing was traditionally opt-in and I wanted to keep it that way. I forgot the two facilities were linked. Since they are, sure - let's merge them.

**Comment by jaywink** [comment]
Technically, the **relays** don't need a statistics.json or NodeInfo. Only **the-federation.info** does. These are two completely separate and different applications.

However, currently the relays require instances to be registered at the-federation.info. This requirement will be lifted at some point, once there is a better way to bootstrap the relays with instances. Maybe something like manual registration to any relay and the relays then share information. But for now, yes, the stats are needed.

Just to clarify, hopefully.

**Comment by zotlabs** [comment]
Understood.

**Comment by git-marijus** [comment]
It kind of works now. I got a post :) It seems though reactions are not being pushed...

**Comment by git-marijus** [comment]
It kind of works now. I got a post :) It seems though reactions are not being pushed...

**Comment by jaywink** [comment]
> It seems though reactions are not being pushed...

None of the three platforms yet push reactions to the relay. The relay itself will relay them to the same place it relayed the target post, once platforms start sending them. The Diaspora issue is diaspora/diaspora#7123

**Comment by git-marijus** [comment]
Thanks for the update @jaywink 

One thing i observed (it might be by accident but i don't think so...): if i register with
```
{
"subscribe": true,
"scope": "tags",
"tags": ["tagone", "tagtwo"]
}
```
i receive some posts. OTOH if i register with:
```
{
"subscribe": true,
"scope": "all"
}
```
i didn't get any allthough i would expect to get more...

**Comment by jaywink** [comment]
Bummer, it seems the relay server is a bit picky here. I wrote the schema [as follows](https://github.com/jaywink/social-relay/blob/master/docs/relays.md#node-subscription-preferences). As you can see `tags` is a required element even though `scope: all` doesn't need the tags list.

While validating schemas is nice and all, I think it's also nicer to show some slack and use whatever info there is. I'll change the poller code to not skip the node if the schema doesn't validate. In the meantime, if you want, you can make yours:

    {
    "subscribe": true,
    "scope": "all",
    "tags": []
    }

Then you should get a lot more posts. But I'll try to fix this in the coming days anyway. Thanks for bringing it to my attention.

**Comment by git-marijus** [comment]
Oups... Thanks will try that...

**Comment by zotlabs** [comment]
Reactions (comments/likes/retractions/whatever) should now be delivered to relays as per the original issue. The code was there already but had an incorrect check for whether we were pushing the activity upstream or downstream.

**Comment by git-marijus** [comment]
I tried with empty tags array included... Still no luck though...

**Comment by jaywink** [comment]
@git-marijus Here are the changes the relay has seen for your node (UTC time):

    2016-10-28 02:09:49,640:DEBUG:poll_pods: Pod: hub.somaton.com, preferences: {"subscribe":true,"scope":"all","tags":[]}
    2016-10-28 08:53:24,629:DEBUG:poll_pods: Pod hub.somaton.com preferences not found, deleting cached data
    2016-10-28 10:00:29,004:DEBUG:poll_pods: Pod: hub.somaton.com, preferences: {"subscribe":true,"scope":"tags","tags":["music","newhere","vienna"]}

During this time, the relay has delivered 157 posts to your node:

    $ grep 'send_document: url=https://hub.somaton.com/receive/public' * | wc -l
    157

What what I can see, the response code is generally 200. So maybe the problem is somewhere else, could you have a look at your logs? Maybe the posts are just not getting processed due to some reason?

@zotlabs awesome! Will try to verify, thanks!

**Comment by git-marijus** [comment]
Thanks for the feedback @jaywink i will dig into it...

**Comment by git-marijus** [comment]
I see... It works...
We don't allow posts if we didn't subscribe to any tags though. If we subscribe to tags we change to scope tags :) 

Need to sort that out... OTOH getting all the traffic will be kind of insane anyway...

**Comment by git-marijus** [comment]
@zotlabs maybe we should implement an admin setting to switch scope between public and tags. What do you think?

**Comment by jaywink** [comment]
As pointed out [in discourse](https://discourse.diasporafoundation.org/t/public-post-federation/264/123), diaspora has no intention to send interactions to the relay. Instead, all received posts will be auto-subscribed to, which will cause source nodes to deliver interactions themselves.

I will make the relay match this and remove interaction forwarding once diaspora starts doing the auto-subscribing. At that point there is no point for Hubzilla to send comments/likes/retractions to the relay either.

Sorry for the extra effort :) Closing.

**Comment by zotlabs** [comment]
Re-opening because the description implies that our code must perform auto-subscription on all received posts and we currently do not. I'm guessing this will be a form of participation message.

**Comment by jaywink** [comment]
👍 

Note, there is no ETA on when Diaspora will start doing this, so I'm not going to remove the relaying of interactions any time soon either. This is just based on discussion with the diaspora team. However, of course, Hubzilla can start auto-subscribing (yes, a participation, I guess) whenever it wants of course, and stop sending interactions to the relay probably at the same time. This should do no harm even if done right away.

**Comment by zotlabs** [comment]
Probably won't happen right away - I'm just scoping out the work in case somebody else wants to tackle it. We don't use D* participations  at all and there's no storage for them so there's a bit of groundwork involved.  We'll probably translate them into relayables and "follow thread" activities and back again; and the way I read it, the server has to follow every thread it pulls from the relay. Currently only channels follow threads so having a server follow a thread is a bit of a mapping issue.
